Mathlib/LinearAlgebra/Multilinear/Basic.lean

```diff
@@ -1603,89 +1603,59 @@ def multilinearCurryRightEquiv :
 
-@[simp]
-theorem uncurrySum_aux_apply
-    (f : MultilinearMap R (fun _ : ι => M') (MultilinearMap R (fun _ : ι' => M') M₂))
-    (u : ι ⊕ ι' → M') : f.uncurrySum u = f (u ∘ Sum.inl) (u ∘ Sum.inr) :=
-  rfl
+attribute [local simp] Sum.update_inl_comp_inl_apply Sum.update_inr_comp_inr_apply
 
-variable (ι ι' R M₂ M')
-
-/-- Linear equivalence between the space of multilinear maps on `∀ i : ι ⊕ ι', M'` and the space
-of multilinear maps on `∀ i : ι, M'` taking values in the space of multilinear maps
-on `∀ i : ι', M'`. -/
-def currySumEquiv :
-    MultilinearMap R (fun _ : ι ⊕ ι' => M') M₂ ≃ₗ[R]
-      MultilinearMap R (fun _ : ι => M') (MultilinearMap R (fun _ : ι' => M') M₂) where
-  toFun := currySum
-  invFun := uncurrySum
-  left_inv f := ext fun u => by simp
-  right_inv f := by
-    ext
-    simp
-  map_add' f g := by
-    ext
-    rfl
-  map_smul' c f := by
+/-- Multilinear maps on `N : (ι ⊕ ι') → Type*` identify to multilinear maps
+from `(fun (i : ι) ↦ N (.inl i))` taking values in the space of
+linear maps on `(fun (i : ι') ↦ N (.inr i))`. -/
+def currySumEquiv : MultilinearMap R N M₂ ≃ₗ[R]
+    MultilinearMap R (fun (i : ι) ↦ N (.inl i))
+      (MultilinearMap R (fun (i : ι') ↦ N (.inr i)) M₂) where
+  toFun f :=
+    { toFun := fun u ↦
+        { toFun := fun v ↦ f (Sum.rec u v)
+          map_update_add' := by intros; letI := Classical.decEq ι; simp
+          map_update_smul' := by intros; letI := Classical.decEq ι; simp }
+      map_update_add' := by intros; letI := Classical.decEq ι'; aesop
+      map_update_smul' := by intros; letI := Classical.decEq ι'; aesop }
+  invFun g :=
+    { toFun := fun u ↦ g (fun i ↦ u (.inl i)) (fun i' ↦ u (.inr i'))
+      map_update_add' := by
+        letI := Classical.decEq ι
+        letI := Classical.decEq ι'
+        rintro _ _ (_ | _) _ _ <;> simp
+      map_update_smul' := by
+        letI := Classical.decEq ι
+        letI := Classical.decEq ι'
+        rintro _ _ (_ | _) _ _ <;> simp }
+  map_add' := by aesop
+  map_smul' := by aesop
+  left_inv f := by
     ext
-    rfl
-
-variable {ι ι' R M₂ M'}
+    simp only [coe_mk]
+    congr
+    ext (_ | _) <;> simp
+  right_inv g := rfl
 
 @[simp]
-theorem coe_currySumEquiv : ⇑(currySumEquiv R ι M₂ M' ι') = currySum :=
-  rfl
+theorem currySumEquiv_apply (f : MultilinearMap R N M₂)
+    (u : (i : ι) → N (Sum.inl i)) (v : (i : ι') → N (Sum.inr i)) :
+    currySumEquiv f u v = f (Sum.rec u v) := rfl
 
--- Porting note: fixed missing letter `y` in name
 @[simp]
-theorem coe_currySumEquiv_symm : ⇑(currySumEquiv R ι M₂ M' ι').symm = uncurrySum :=
-  rfl
+theorem currySumEquiv_symm_apply
+    (g : MultilinearMap R (fun (i : ι) ↦ N (.inl i))
+      (MultilinearMap R (fun (i : ι') ↦ N (.inr i)) M₂)) (u) :
+    currySumEquiv.symm g u =
+      g (fun i ↦ u (.inl i)) (fun i' ↦ u (.inr i')) := rfl
+
+@[deprecated  (since := "2024-11-01")]
+alias currySum := currySumEquiv
+@[deprecated  (since := "2024-11-01")]
+alias uncurrySum := currySumEquiv
```

Comment:
I don't think that `deprecated alias` is appropriate here, since this completely changes the type of these declarations (there is a missing `.symm`!)

My preference would be to keep these as the simp-normal spellings, just as `curry` and `uncurry` are simp-normal while `Equiv.curry` is a bundling for convenience when needed.

Reply:
Thanks for pointing this out. I have now use a deprecated `def` instead.

I see little reasons to use `currySum/uncurrySum` (+ `left_inv/right_inv` lemmas) instead of a single definition `currySumEquiv` (which shall be very rarely used anyway).

Reply:
One argument for keeping these would be consistency with `ContinuousMultilinearMap.currySum` etc. Maybe @sgouezel and @urkud have opinions on whether the not-fully-bundled versions are worth keeping around.

Reply:
The not fully bundled versions are much lighter to use in practice, so we tend to use these most of the time, and only switch to the bundled version when its properties are important.

Reply:
If `(un)currySum` are to be the simp NF, in principle, we should also add lemmas expressing these plain maps are linear, etc, all of which is taken care of using a `LinearEquiv`.

Reply:
The only reasonable compromise I may consider would be to define the basic `currySum/uncurrySum` as abbreviations referring to `currySumEquiv`.

Reply:
Can we at split this PR into:

* Add the new dependent version of `currySum`/`uncurrySum`
* Refactor the simp-normal form

I'd prefer not to have to do the former and the latter at the same time, and the status quo is to use `currySum` and `uncurrySum` as simp-normal.

Reply:
I do not see the point of splitting this PR into two.

What I am saying is that currently, we do not have basic API lemmas like `currySum_add`, `currySum_smul`, etc, not even `currySum_uncurrySum`, which we should have if the API is meant to involve the basic definitions `(un)currySum`. If everything is defined in terms of `currySumEquiv`, then we have all of these through the `LinearEquiv` API. If some people are not happy using `currySumEquiv`, they may use basic `currySum` and `uncurrySum` if these are defined as abbreviations as I am suggesting above.

Reply:
I agree those lemmas are missing; my preference would be to add at least the `add` and `smul` versions (which we already prove manually anyway, so this is just a job of extracting the proofs to lemmas), and let people rewrite by `currySumEquiv_apply` if they need access to the full API. I think this is fairly consistent with how things are handled across mathlib, even if it may not be the perfect choice.

Reply:
My preference is not to duplicate the generic `map_add` and `map_smul`.

Reply:
I have played a lot with bundled and unbundled versions of currying maps for continuous multilinear maps, when defining higher derivatives and proving their properties (see for instance the definition https://github.com/leanprover-community/mathlib4/blob/7b8edbb98887af54a4cb8e8f76fa279960233c22/Mathlib/Analysis/Calculus/ContDiff/FTaylorSeries.lean#L344 and basically all the corresponding file). And if I remember correctly there were sometimes weird elaboration issues when using the bundled versions, and things were slower because the types are complicated (you need to understand the algebraic / topological properties of continuous linear maps into continuous mutlilinear maps and friends whenever you write down the equiv, and this is putting typeclass inference to its knees). This is why I used the unbundled version whenever possible, and only switched to the bundled version if it was necessary -- the presence of `set_option maxSynthPendingDepth 2` in this file is an indication of what goes wrong sometimes. So, I really don't think it's a wise move to go with the bundled version as the standard one.